### PR TITLE
Quantity searches update to consider implicit ranges #257

### PR DESCRIPTION
**Pull request by prb112**
- Remove reference to PARAMETERS_GTT which is not used and also removed
insert method from the interface and impl
- Refactor JDBCQueryBuilder to move QuantityParameter processing into a
Specific Utility related to the processing
- Add QuantityParmBehaviorUtil
- Add tests for QuantityParmBehaviorUtil and the various methods and
prefixes
- Add QuantityParmBehaviorUtilTest to testng.xml
- Reset NumberParmBehaviorUtilTest to FINE (avoids needless logs at
INFO)

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/QuantityParmBehaviorUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+        List<Object> actualBindVariables = new ArrayList<>();
+
+        QuantityParmBehaviorUtil behavior = new QuantityParmBehaviorUtil();
+        behavior.executeBehavior(actualWhereClauseSegment, queryParm, actualBindVariables,
+                tableAlias, generateDao(sendNull));
+
+        if (log.isLoggable(LOG_LEVEL)) {
+            log.info("whereClauseSegment -> " + actualWhereClauseSegment.toString());
+            log.info("bind variables -> " + actualBindVariables);
+        }
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/QuantityParmBehaviorUtilTest.java]: i havn't seen this pattern used in our codebase for INFO-level messages...are you sure we need to guard these?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 2]
```diff
@@ -38,42 +38,41 @@ public void testSearchQuantity_Quantity() throws Exception {
...
     public void testSearchQuantity_Quantity_chained() throws Exception {
         assertSearchReturnsComposition("subject:Basic.Quantity", "25|http://unitsofmeasure.org|s");
         assertSearchReturnsComposition("subject:Basic.Quantity", "25||s");
-        
+
         // DSTU2 does not say if this is allowed or not, but we do not support it.
         // In more recent versions, they clarified that it should work:  https://build.fhir.org/search.html#quantity
-//        assertSearchReturnsComposition("Quantity", "25");
-        
+        //  assertSearchReturnsComposition("Quantity", "25");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Whats the plan for addressing this one?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 3]
```diff
@@ -38,42 +38,41 @@ public void testSearchQuantity_Quantity() throws Exception {
...
+
         // I think this should return the resource but it currently doesn't.
         // https://gforge.hl7.org/gf/project/fhir/tracker/?action=TrackerItemEdit&tracker_item_id=19597
-//        assertSearchReturnsComposition("Quantity", "25||sec");
+        // assertSearchReturnsComposition("Quantity", "25||sec");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Does this test work now?  If not, whats the plan to address this one?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 4]
```diff
@@ -38,42 +38,41 @@ public void testSearchQuantity_Quantity() throws Exception {
...
-        
+
         // DSTU2 does not say if this is allowed or not, but we do not support it.
         // In more recent versions, they clarified that it should work:  https://build.fhir.org/search.html#quantity
-//        assertSearchReturnsComposition("Quantity", "25");
-        
+        //  assertSearchReturnsComposition("Quantity", "25");
+
         // I think this should return the resource but it currently doesn't.
         // https://gforge.hl7.org/gf/project/fhir/tracker/?action=TrackerItemEdit&tracker_item_id=19597
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: HL7 moved from "GForge" to JIRA.  The link should now be c
```suggestion
        // https://jira.hl7.org/browse/FHIR-19597
```

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 5]
```diff
@@ -170,163 +199,200 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
         // spec isn't clear about whether quantities with no unit should be indexed
         // but since we require the unit while searching, it doesn't really matter
         assertSearchDoesntReturnSavedResource("Quantity-noCodeOrUnit", "1||eq");
     }
-    
+
     /***
-     * FHIR Server does not yet use quantity comparator to calculate search results. *
+     * FHIR Server does not yet use quantity comparator to calculate search results.
+     * *
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: this line can be deleted

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: you know I didn't even hit that one.  Let me try locally once you are done with your review.  I'll get it in the next update

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 6]
```diff
@@ -170,163 +199,200 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+
     @Test
     public void testSearchQuantity_Quantity_GreaterThan() throws Exception {
         // Later versions of the spec indicate that there is an implicit precision 
         // of .5 of the next least significant digit.  We don't support that now, but 
         // lets use numbers far enough away that it won't matter.
         assertSearchDoesntReturnSavedResource("Quantity-greaterThan", "2||gt");
-//        assertSearchReturnsSavedResource("Quantity-greaterThan", "4||gt");
-        
+        //        assertSearchReturnsSavedResource("Quantity-greaterThan", "4||gt");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: can this be uncommented?

Comment (lmsurpre) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: or is it commented still because we don't consider Quantity comparators?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 7]
```diff
@@ -170,163 +199,200 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        assertSearchReturnsSavedResource("Quantity-lessThan", "lt4||lt"); // < 3 may be < 4 
         assertSearchDoesntReturnSavedResource("Quantity-lessThan", "gt4||lt"); // < 3 is not > 4
     }
-    
+
     @Test
     public void testSearchQuantity_Quantity_GreaterThan() throws Exception {
         // Later versions of the spec indicate that there is an implicit precision 
         // of .5 of the next least significant digit.  We don't support that now, but 
         // lets use numbers far enough away that it won't matter.
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: we can remove this comment now that we support significant digits

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 8]
```diff
@@ -170,163 +199,200 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
         assertSearchDoesntReturnSavedResource("Quantity-greaterThan", "2||gt");
-//        assertSearchReturnsSavedResource("Quantity-greaterThan", "4||gt");
-        
+        //        assertSearchReturnsSavedResource("Quantity-greaterThan", "4||gt");
+
         // With implicit ranges, 3 (+/-0.5) actually might be > 3
-//      assertSearchDoesntReturnSavedResource("Quantity-greaterThan", "3||gt");
-        
+        //      assertSearchDoesntReturnSavedResource("Quantity-greaterThan", "3||gt");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: can this be uncommented?

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/QuantityParmBehaviorUtilTest.java (hunk 1 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/QuantityParmBehaviorUtilTest.java]: Fixed

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 6 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Right no - comparators.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 8 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Discussed offline.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 7 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Done

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I think I resolved.  please double check (leaving unresolved)

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 3 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I think I resolved.  please double check (leaving unresolved)

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 9]
```diff
@@ -82,80 +75,110 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
+    public void testSearchQuantity_Quantity_withPrefix_GT() throws Exception {
         assertSearchReturnsSavedResource("Quantity", "gt24|http://unitsofmeasure.org|s");
         assertSearchReturnsSavedResource("Quantity", "gt24.4999||s");
         assertSearchReturnsSavedResource("Quantity", "gt24.5||s");
-        assertSearchDoesntReturnSavedResource("Quantity", "gt25||s");
+
+        // For the following test, as the upper bound is now the lower bound of the range
+        // the following must use a lower value of the range/precision 
+        // To keep from hitting, we have to make it more precise. 
+        assertSearchDoesntReturnSavedResource("Quantity", "gt25.05||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: i don't get it.  shouldn't the presence of `gt` prefix make us treat `gt25` as if it were `gt25.0000000000...` ?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 10]
```diff
@@ -82,80 +75,110 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
+
+    @Test
+    public void testSearchQuantity_Quantity_withPrefix_LE() throws Exception {
         assertSearchDoesntReturnSavedResource("Quantity", "le24|http://unitsofmeasure.org|s");
         assertSearchDoesntReturnSavedResource("Quantity", "le24.4999||s");
         assertSearchDoesntReturnSavedResource("Quantity", "le24.5||s");
-        assertSearchReturnsSavedResource("Quantity", "le25||s");
+
+        // We have to make the following test more precise since the implied range is used. 
+        assertSearchReturnsSavedResource("Quantity", "le25.01||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: i don't get it. shouldn't the presence of gt prefix make us treat le25 as if it were le25.0000000000... ?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 11]
```diff
@@ -82,80 +75,110 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
+        assertSearchDoesntReturnSavedResource("Quantity", "eb24.4999||s");
         assertSearchReturnsSavedResource("Quantity", "eb25.5||s");
         assertSearchReturnsSavedResource("Quantity", "eb26|http://unitsofmeasure.org|s");
     }
-    
+
     @Test
     public void testSearchQuantity_Quantity_withPrefixes_chained() throws Exception {
-        assertSearchReturnsComposition("subject:Basic.Quantity", "lt26|http://unitsofmeasure.org|s");
+        assertSearchReturnsComposition("subject:Basic.Quantity", "lt26.0|http://unitsofmeasure.org|s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: extra precision shouldn't be needed due to `lt` prefix

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 12]
```diff
@@ -82,80 +75,110 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
         assertSearchReturnsComposition("subject:Basic.Quantity", "gt24|http://unitsofmeasure.org|s");
         assertSearchReturnsComposition("subject:Basic.Quantity", "le26|http://unitsofmeasure.org|s");
-        assertSearchReturnsComposition("subject:Basic.Quantity", "le25|http://unitsofmeasure.org|s");
+        assertSearchReturnsComposition("subject:Basic.Quantity", "le25.02|http://unitsofmeasure.org|s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: extra precision shouldn't be needed due to `le` prefix

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 13]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
-        // of .5 of the next least significant digit.  We don't support that now, but 
-        // lets use numbers far enough away that it won't matter.
-//        assertSearchReturnsSavedResource("Quantity-lessThan", "2||lt");
+        // For more details, please see Quantity search does not consider the quantity comparator field #493 https://github.com/IBM/FHIR/issues/493
+        // assertSearchReturnsSavedResource("Quantity-lessThan", "2||lt");
         assertSearchDoesntReturnSavedResource("Quantity-lessThan", "4||lt");
-        
+
+        // For more details, please see Quantity search does not consider the quantity comparator field #493 https://github.com/IBM/FHIR/issues/493
         // With implicit ranges, 3 (+/-0.5) actually might be < 3
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: This comment should be removed if we remove the test it was commenting, which was:
`assertSearchDoesntReturnSavedResource("Quantity-lessThan", "3||lt");`

However, I don't think this test should have been removed (see next comment)

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 14]
```diff
@@ -82,74 +75,104 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
+
+    @Test
+    public void testSearchQuantity_Quantity_withPrefix_GT() throws Exception {
         assertSearchReturnsSavedResource("Quantity", "gt24|http://unitsofmeasure.org|s");
         assertSearchReturnsSavedResource("Quantity", "gt24.4999||s");
         assertSearchReturnsSavedResource("Quantity", "gt24.5||s");
+
+        // For the following test, as the upper bound is now the lower bound of the range
+        // the following must use a lower value of the range/precision 
+        // To keep from hitting, we have to make it more precise.
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: i think this comment should be removed now

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 15]
```diff
@@ -82,74 +75,104 @@ public void testSearchQuantity_Quantity_revinclude() throws Exception {
...
+
+    }
+
+    @Test
+    public void testSearchQuantity_Quantity_withPrefix_LE() throws Exception {
         assertSearchDoesntReturnSavedResource("Quantity", "le24|http://unitsofmeasure.org|s");
         assertSearchDoesntReturnSavedResource("Quantity", "le24.4999||s");
         assertSearchDoesntReturnSavedResource("Quantity", "le24.5||s");
+
+        // We have to make the following test more precise since the implied range is used.
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: remove this comment

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 16]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+
+    @Test
+    public void testSearchQuantity_Range_LT() throws Exception {
         assertSearchDoesntReturnSavedResource("Range", "lt4||s");
         assertSearchDoesntReturnSavedResource("Range", "lt5||s");
-        assertSearchReturnsSavedResource("Range", "lt10||s");
+
+        // Lower Bound is 9.5 for 11, therefore adding precision.
+        // And this is still a RANGE search.
+        assertSearchReturnsSavedResource("Range", "lt10.05||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I suggest restoring this to `lt10||s` ... this should still return the saved resource because the range `lt10` overlaps with the range of the value which is `5 - 10`

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 17]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
-        
+    }
+
+    @Test
+    public void testSearchQuantity_Range_EB() throws Exception {
         assertSearchDoesntReturnSavedResource("Range", "eb4||s");
         assertSearchDoesntReturnSavedResource("Range", "eb5||s");
-        assertSearchDoesntReturnSavedResource("Range", "eb10||s");
+        // We use the range, so we actually return the value here. 
+        assertSearchReturnsSavedResource("Range", "eb10||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: Probably not a big deal, but I disagree with this change.  https://www.hl7.org/fhir/datatypes.html#Range says
> The low and the high values are inclusive and are assumed to have arbitrarily high precision; e.g. the range 1.5 to 2.5 includes 1.50, and 2.50 but not 1.49 or 2.51

Because 5-10 include 10.000... and eb10 includes 10.000 (both with arbitrarilly high precision) I think it should return empty.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 18]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
         assertSearchReturnsSavedResource("Range", "eb11||s");
     }
+
+    @Test
+    public void testSearchQuantity_Range_SA() throws Exception {
+        assertSearchReturnsSavedResource("Range", "sa4||s");
+        assertSearchDoesntReturnSavedResource("Range", "sa10||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: we're missing the lower bound test here. i suggestion adding a `sa5||s` between these two lines

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 19]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        assertSearchDoesntReturnSavedResource("Range", "sa10.0||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: since `sa` prefix should make it into arbitrarilly precision, this should be the same as the previous line.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 20]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        assertSearchDoesntReturnSavedResource("Range", "sa11||s");
+    }
     
+    @Test
+    public void testSearchQuantity_Range_GE() throws Exception {
+        assertSearchReturnsSavedResource("Range", "ge4||s");
+        assertSearchReturnsSavedResource("Range", "ge5||s");
+        // We're using the lowerbound to trigger the range search
+        assertSearchDoesntReturnSavedResource("Range", "ge10||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I think it *should* return the resource because the Range datatype upper bound is inclusive and so technically the ranges overlap

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 21]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        assertSearchDoesntReturnSavedResource("Range", "ge11||s");
+    }
+
+    @Test
+    public void testSearchQuantity_Range_LE() throws Exception {
+        assertSearchDoesntReturnSavedResource("Range", "le4||s");
+        assertSearchDoesntReturnSavedResource("Range", "le5.01||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: we're missing the exact edge case (the one thing to distingush `le` from `lt`).   I suggest either turning this line into `le5||s` or introducing that above this line.
```suggestion
        assertSearchReturnsSavedResource("Range", "le5||s");
```

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 22]
```diff
@@ -170,163 +193,201 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        // the upper bound is actually higher than the valueRange we inserted. 
+        assertSearchDoesntReturnSavedResource("Range", "le10||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I disagree with this one too.  The target range is 5-10 which is le10....`le` should be looking for overlapping ranges and these range definitely overlap.

Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: great find. you are right. it was using lowerbound, it should have been using upperbound.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 21 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: in next commit.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 18 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: check

Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I have changed sa/eb to reflect the above examples.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 20 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: fixed, it was passing in the wrong bound.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 23]
```diff
@@ -170,163 +187,195 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+        assertSearchReturnsSavedResource("Range", "10||s");
+        assertSearchDoesntReturnSavedResource("Range", "11||s");
+    }
+
+    @Test
+    public void testSearchQuantity_Range_LT() throws Exception {
         assertSearchDoesntReturnSavedResource("Range", "lt4||s");
         assertSearchDoesntReturnSavedResource("Range", "lt5||s");
-        assertSearchReturnsSavedResource("Range", "lt10||s");
+        assertSearchDoesntReturnSavedResource("Range", "lt10||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: this *should* return the resource because `lt10` overlaps `5-10`

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 24]
```diff
@@ -170,163 +187,195 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
         assertSearchReturnsSavedResource("Range", "lt11||s");
-        
+    }
+
+    @Test
+    public void testSearchQuantity_Range_GT() throws Exception {
         assertSearchReturnsSavedResource("Range", "gt4||s");
-        assertSearchReturnsSavedResource("Range", "gt5||s");
+        assertSearchDoesntReturnSavedResource("Range", "gt5||s");
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: this *should* return the resource because `gt5` overlaps `5-10`

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 25]
```diff
@@ -170,163 +187,195 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
+
+        assertSearchDoesntReturnSavedResource("Quantity-lessThanOrEqual", "lt2||lte");      // <= 3 may be < 2
+        assertSearchReturnsSavedResource("Quantity-lessThanOrEqual", "gt2||lte"); // <= 3 may be > 2
+        assertSearchReturnsSavedResource("Quantity-lessThanOrEqual", "lt4||lte"); // <= 3 may be < 4 
         assertSearchDoesntReturnSavedResource("Quantity-lessThanOrEqual", "gt4||lte"); // <= 3 is not > 4
-        assertSearchReturnsSavedResource("Quantity-lessThanOrEqual", "le3||lte");      // <= 3 is <= 3
-        assertSearchReturnsSavedResource("Quantity-lessThanOrEqual", "ge3||lte");      // <= 3 may be >= 3
+        
+        // As we have added implict ranges to the prefix processing, we need to add precision 
+        // >= 3 may be <= 3 uses precision and bounding for the range.
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I suggest removing this.  At least the first line and probably the second as well.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 26]
```diff
@@ -170,163 +187,195 @@ public void testSearchQuantity_Quantity_NoDisplayUnit() throws Exception {
...
-        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "lt4||gte");      // >= 3 may be < 4 
-//        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "gt4||gte");      // >= 3 may be > 4
-        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "le3||gte");         // >= 3 may be <= 3
-        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "ge3||gte");         // >= 3 is >= 3
+        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "gt2||gte"); // >= 3 may be > 2
+        assertSearchReturnsSavedResource("Quantity-greaterThanOrEqual", "lt4||gte"); // >= 3 may be < 4 
+        assertSearchDoesntReturnSavedResource("Quantity-greaterThanOrEqual", "gt4||gte");      // >= 3 may be > 4
+        
+        // As we have added implict ranges to the prefix processing, we need to add precision 
+        // >= 3 may be <= 3 uses precision and bounding for the range.
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java]: I suggest removing this.  At least the first line and probably the second as well.
